Design for integrating kopia with kanister (#1482)
kanisterio/kanister: design/kanister-kopia-integration.md

- To authorize access, a list of server usernames and passwords must be added prior to starting the server.
- The server also uses TLS certificates to secure incoming connections to it.
- Kanister users can configure a repository via a newly added Custom Resource called `RepositoryServer` as described in the following section.

#### Custom Resource Definition

This design proposes a new Custom Resource Definition (CRD) named `RepositoryServer`, to represent a Kopia repository server.
It is a namespace-scoped resource, owned and managed by a new `RepositoryServerController` controller in the `kanister` namespace.
As mentioned above, the CRD offers a set of parameters to configure a Kopia repository server.

To limit the new controller's RBAC scope, all the `RepositoryServer` resources are created in the `kanister` namespace.

A sample `RepositoryServer` resource created to interact with the Kopia repository of workloads running in the `monitoring` namespace looks like this:

```yaml
apiVersion: cr.kanister.io/v1alpha1
kind: RepositoryServer
metadata:
   name: repository-monitoring
   namespace: kanister
   labels:
      repo.kanister.io/target-namespace: monitoring
spec:

>    location:

Author: As mentioned in [this comment](https://github.com/kanisterio/kanister/pull/1645#discussion_r979895726), can we rename the `location` field here?

The Location struct for Profile CR [in pkg/apis/cr/v1alpha1/types.go](https://github.com/kanisterio/kanister/blob/dab83e87b16e7eeb73645ed0768450fd978ee2b7/pkg/apis/cr/v1alpha1/types.go#L255) is conflicting with this location field here

Reviewer: @shlokchaudhari9 Yes, I think `storage` might work here as well. I'll change it 👍🏼

      # required: storage location info
      secret:
         name: location
         namespace: kanister
      # required: creds to access the location
      credentialSecret:
         name: loc-creds

>          namespace: kanister

Reviewer 2: We should use the `secretRef` suffix, following the K8s API convention. E.g., see [here](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#csipersistentvolumesource-v1-core). In the code, this should be back by the [`corev1.SecretReference`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#secretreference-v1-core) struct.

   repository:
      # repository must be created manually before creating server CR
      # required: path for the repository - will be relative sub path
      # within the path prefix specified in the location
      rootPath: /repo/monitoring/
      # required: password to access the repository
      passwordSecret:
         name: repository-monitoring-password
         namespace: kanister
      # required: these values will be used by the repository server controller
      # for the connection from the server to the repository. Kopia will use
      # these values to track and control operations
      username: kanisterAdmin

>       hostname: monitoring.app

Reviewer 2: When and how do these get created? They can take any values, right?

   server:
      # required: server admin details
      adminSecret:
         name: repository-monitoring-server-admin
         namespace: kanister
      # required: TLS certificate required for secure communication between the Kopia client and server
      tlsSecret:
         name: repository-monitoring-server-tls
         namespace: kanister
      # required: repository users list
      userAccessSecret:
         name: repository-monitoring-server-access
         namespace: kanister
      # required: selector for network policy required to enable cross-namespace access to the server

>       labelSelector:

Reviewer 2: Can we be more explicit about this being selectors for network policy, and not e.g., `Service`? How about, following the example in the network policy [doc](https://kubernetes.io/docs/concepts/services-networking/network-policies/#networkpolicy-resource), something like:

```yaml
server:
    networkPolicy:
        namespaceSelector:
            matchLabels:
                project: myproject
        podSelector:
            matchLabels:
                role: frontend
```

Author: I agree with @ihcsim here. @pavannd1 we should make the `labelSelector` field more descriptive. Users might get confused about the type of these labels. They might question whether the labels here are pod-selectors or namespace-selectors. Better we explicitly ask them.

Author: Also makes it easier for controller to define the ingress rule in `NetworkPolicy`. Rather than making assumptions about the kind of selector these labels would be, controller would directly use these selectors while creating the `NetworkPolicy`.

         matchLabels:
         - key1: value1
         - key2: value2
status:
   conditions:
   - lastTransitionTime: "2022-08-20T09:48:36Z"
     lastUpdateTime: "2022-08-20T09:48:36Z"
     status: "True"
     type: RepositoryServerReady
   serverInfo:
      podName: "repository-monitoring-pod-a1b2c3"
      networkPolicyName: "repository-monitoring-np-d4e5f6"
      serviceName: "repository-monitoring-svc-g7h8i9"
```

The required `spec.location.secret` refers to a `Secret` resource storing location-related sensitive data. This secret is provided by the user.
For example,

```yaml
apiVersion: v1
kind: Secret
metadata:
   name: location
   namespace: kanister
   labels:
      repo.kanister.io/target-namespace: monitoring
type: Opaque
data:
   # required: supported values are s3, gcs, and azure
   type: s3
   # required
   bucket: my-bucket
   # optional: specified in case of S3-compatible stores
   endpoint: https://foo.example.com
   # optional: used as a sub path in the bucket for all backups
   path: kanister/backups
   # required, if supported by the provider
   region: us-west-1
   # optional: if set to true, do not verify SSL cert.
   # Default, when omitted, is false

>    skipSSLVerify: false

Reviewer 2: We should document all the supported keys. to help the user to put this secret together. AFAICT, the only place where this is documented now is in the `kopia` CLI [code](https://github.com/kopia/kopia/blob/05e729a7858a6e86cb48ba29fb53cb6045efce2b/cli/storage_s3.go#L19-L27).

is used to determine the labels used in the `podSelector` of the `NetworkPolicy` resource that controls the ingress
traffic to the repository server.

The `status` subresource provides conditions and status information to ensure that the controller does not attempt to re-create
the repository server during restart.

#### Repository Server Lifecycle

When a `RepositoryServer` resource is created, the `RepositoryController` controller responds by creating a `Pod`, `Service`,
and a `NetworkPolicy` as mentioned above.
These resources are cleaned up when the `RepositoryServer` resource is deleted.

#### Server-side Setup

Once the pod is running, the controller executes a set of Kopia CLI commands as follows.

> 📝 The examples in this section use S3 for illustration purposes.

- Establish a connection to the Kopia repository

```sh
kopia repository connect s3 \
   --bucket=my-bucket \
   --access-key=<redacted> \
   --secret-access-key=<redacted> \
   [--endpoint=https://foo.example.com \]
   [--prefix=my-prefix \]
   [--region=us-west-1 \]
   --password=<redacted> \
   --override-hostname=<hostname> \

>    --override-username=<username>

Reviewer 2: Do we know if this is still necessary or is `kopia server start` smart enough to figure it out if we add all the repo info into the `repo.config` file?

Reviewer: A connect operation is required even if the configuration is added to a file.
We are exploring ways to utilize the `kopia repository connect from-config` option provided by Kopia to supply all the information via a single config file.

- Register the set of users that have access to the server

```sh
kopia server user add <username>@<hostname> --user-password=<redacted>
```

The username, hostname, and password will be picked up from the `spec.server.accessSecret` property.

- Refresh the server process to enable the newly added users

```sh
kopia server refresh --server-cert-fingerprint=<redacted> \
   --address=0.0.0.0:51515 \
   --server-username=<redacted> \
   --server-password=<redacted>
```

> 📝 The secrets provided in the `RepositoryServer` resource are mounted via the pod's `spec.volumes` API.

#### Client-side Setup

The Kopia server is fronted by a K8s `Service` resource. Data mover clients will connect to it using the equivalence of:

```sh
kopia repository connect server \
   --url=https://<service-name> \
   --config-file=/run/kopia/repo.config \
   --server-cert-fingerprint=<redacted> \
   --override-username=<username> \
   --override-hostname=<hostname> \
   --password=<redacted>
```

The `<service-name>` is the Kopia server's `Service` resource name.
This will be auto-generated by the repository controller and provided via the `status` subresource of the `RepositoryServer` resource.

> The `server-cert-fingerprint` is derived from the TLS certificates provided during the creation of the server resource.

Reviewer 2: It's ok to add the fingerprint to the `status` too, since it's public.

# Integrating Kopia with Kanister

> This document proposes all the changes required within Kanister to better support the use of Kopia backup/restore tool.

Reviewer 2: ```suggestion
This document proposes all the changes required within Kanister to use [Kopia](https://kopia.io/) as the primary backup and restore tool.
```

## Introducing Kopia

> [Kopia](https://kopia.io/) is a powerful, cross-platform tool for managing encrypted backups in the cloud.

Reviewer 2: ```suggestion
Kopia is a powerful, cross-platform tool for managing encrypted backups in the cloud.
```

It provides fast and secure backups, using compression, data deduplication and client-side end-to-end encryption.
It supports a variety of backup storage targets, including object stores, which allows users to choose the storage provider that better addresses their needs.

> It is a lock-free system that allows for concurrent multi-client operations including garbage collection.

Reviewer 2: ```suggestion
It is a lock-free system that allows for concurrent multi-client operations including garbage collection.

To explore other features of Kopia, see its [documentation](https://kopia.io/docs/features/).
```

## Goal

Kopia can focus on its core data export and transformation as part of various Kanister functions.

> While, Kanister continues to execute these functions from it's application-centric data protection workflows called Blueprints.

Reviewer 2: I am not sure what this means and how this benefits Kanister users and downstream applications. I think this section should describe what problems we are trying to solve, and what is lacking in the current approach.

> ##  Proposed Work

Reviewer 2: s/Proposed Work/Scope

### Kanister Functions 

We already have a rich repository of Kanister functions present on path `pkg/function` that enable application-level data protection in various use cases.
Few of these functions need to be re-worked to use Kopia. For starters, these functions are;

1. BackupData
2. BackupDataAll
3. BackupDataStats
4. CopyVolumeData
5. DeleteData
6. DeleteDataAll
7. RestoreData
8. RestoreDataAll

> Please note that the functions stated above will only be **refactored** to use Kopia underneath.

Reviewer 3: Will we be able to use something besides Kopia or will this make Kopia the only mechanism supporting these functions?

Author: We can continue to maintain separate versions of these Kanister Functions. For ex; `v.0.0.0` of BackupData (or any other Function) will continue to follow restic workflow and a version like `v1.0.0-alpha` could be implemented using Kopia.

The reason I mentioned "refactor" is because in a previous discussion with @pavannd1, we thought one of the end goal here could be to completely replace the Restic based Kanister Functions with the one's that use Kopia. But if that's still unclear, we can decide on it at a later point since function versions are easily configurable by downstream users.

Reviewer 4: Can we design the data-mover interface in such a way that in the future, if we want to switch to another tool it will be easier than refactoring all the functions? I understand this could be out of the current scope but just wondering.

There should not be any changes with respect to the objective of these functions.
The motive, arguments and usage for each function stays intact as defined on Kanister docs https://docs.kanister.io/functions.html#existing-functions

### Kanister Functions that use Kopia API Server 

We should also create Kanister functions that leverage the use of [Kopia in it's server mode](https://kopia.io/docs/repository-server/).
These Kanister functions would act like Kopia clients that securely proxy access repository storage without exposing sensitive storage credentials.

Following is a list of few such functions;

1. BackupDataToServer - To perform backup via the KopiaAPIserver

> 2. RestoreDataFromServer - To restore from the KopiaAPIServer

Reviewer 3: Using a server versus the local integration should be configuration driven, not be different functions.

Author: This makes a lot of sense.... Noting it on the design doc 👍🏼

Reviewer 2: Why do we need both? Why not just use the client/server setup only?

Reviewer: I think the idea is to always use the client/server setup so we can eliminate the direct-repo functions 👍🏼

# Integrating Kopia with Kanister

This document proposes all the changes required within Kanister to encourage the use of Kopia backup/restore tool.

## Introducing Kopia

Kopia is a simple, cross-platform tool for managing encrypted backups in the cloud.
It provides fast, incremental backups, secure, client-side end-to-end encryption, compression and data deduplication.
It is one of the fastest backup/restore tool in the field and with a pluggable backend engine, it allows user to decide which storage the backups should go to.
It is a lock-free system that, unlike other similar tools, allows for concurrent multi-client operations including garbage collection.

To learn more about Kopia and it's features, visit https://kopia.io/

## Goal

> Kopia can focus on its core data transformation as part of various Kanister functions.

Author: Yes. Also, data export.

While, Kanister takes care of executing these functions from it's application-centric data protection workflows called Blueprints.

##  Proposed Work

### Kanister Functions 

> We already have a rich repository of Kanister functions present on path `pkg/function` that enable application-level data protection in various use-cases.

Reviewer 5: ```suggestion
We already have a rich repository of Kanister functions present on path `pkg/function` that enable application-level data protection in various use cases.
```

Few of these functions need to be re-worked to use Kopia. For starters, these functions are;

1. BackupData
2. BackupDataAll
3. BackupDataStats
4. CopyVolumeData
5. DeleteData
6. DeleteDataAll
7. RestoreData
8. RestoreDataAll

Please note that the functions stated above will only be **refactored** to use Kopia underneath.
There should not be any changes with respect to the objective of these functions.
The motive, arguments and usage for each function stays intact as defined on Kanister docs https://docs.kanister.io/functions.html#existing-functions

### Kanister Functions that use Kopia API Server 

We should also create Kanister functions that leverage the use of [Kopia in it's server mode](https://kopia.io/docs/repository-server/).
These Kanister functions would act like Kopia clients that securely proxy access repository storage without exposing sensitive storage credentials.

Following is a list of few such functions;

1. BackupDataToServer - To perform backup via the KopiaAPIserver
2. RestoreDataFromServer - To restore from the KopiaAPIServer

### Limitations to using Kopia API Server

> Before starting Kopia API Server, it demands a list of usernames and passwords that are allowed access.

Reviewer 5: ```suggestion
Before starting Kopia API Server, it requires the creation of users that are allowed access.
```

> Before starting Kopia API Server, it demands a list of usernames and passwords that are allowed access.

Reviewer 5: ```suggestion
Before starting Kopia API Server, it requires the creation of user that are allowed access with their corresponding passwords.
```
```suggestion
Before starting Kopia API Server, it demands a list of usernames and passwords that are allowed access.
```

We can make use of the application namespace ID to generate usernames and a random alphanumeric encryption key to create passwords.
These would be stored in a secret for later use. But;

> - How should we store the server username and server password securely?

Reviewer 2: Do these secrets have to be generated internally by the server? Or can it be:

* generated by a human user and store that in a `Secret` resource, as part of the installation process (default use case)
* generated and managed by some KMS solutions + [CSI secret store](https://secrets-store-csi-driver.sigs.k8s.io/) (advance, prod-ready use case)

The idea is to use the K8s `Secret` API as the contract between the user, the server and any downstream clients that may need to share these credentials. The responsibility of securing the at-rest storage falls on the user or the KMS solution.

Author: I really like the idea of having a secret store to manage these secrets (2nd use case)

Reviewer 3: Using a server vs local should be configuration driven, not hard-coded into the functions

> - At which point in Kanister, should Kopia API Server be started?

Reviewer 2: Instead of starting this on-demand, can it be deployed as a separate, required workload? Feels like a lightweight server.

Reviewer 5: In the first iteration, it may make sense to have the server running as a requirement. Then functionality to launch and manage the server can be added to Kanister.

> - How could this Kopia API Server be disabled in case downstream Kanister consumers bring their own Kopia API Server?

Reviewer 2: If we can deploy the server as a separate workload, we should be able to add a Helm option to disable the deployment of this workload, and another option to configure the server endpoint that Kanister controller calls. Will that work?

Author: That should work assuming Kopia API server has the ability to connect and disconnect underlying storage whenever required. cc: @pavannd1

Reviewer: Maybe Kopia documentation has this information but we need to experiment and see if we can connect/disconnect to/from the repository while the server is running. If yes, we can start the server pod/container based on Kanister installation config.

Reviewer 2: @shlokchaudhari9 sounds like we have some stress test for Kopia that can help determine throughput and latency. This will give us some controlled benchmark info on scalability i.e. how many backup requests can one Kopia pod support, how many backup requests should trigger HPA etc. Good to include this info in the design doc. @pavannd1 knows where to find these tests.

Author: @ihcsim Okay. I'll go through these tests and add details on doc for server scalability.

   the required Kopia repository and the repository server before executing the
   actions.
- After setting up the repository and the repository server, users can follow
   the normal workflow to execute actions from the v2 Blueprints. To use the
   new versions of the Kanister Data Functions, users must specify the version
   of the function via the ActionSet Action's `preferredVersion` field. This
   field in an Action is applied to all the Phases in it. If a particular
   Kanister function is not registered with this version, Kanister will fall
   back to the default version of the function.

## Detailed Design

### Kanister Data Functions

- Kanister allows mutliple versions of Functions to be registered with the
   controller.
- Existing Functions are registered with the default `v0.0.0` version. Find
   more information
   [here](https://docs.kanister.io/functions.html#existing-functions).
- The following Data Functions will be registered with a second
   version `v1.0.0-alpha`:

   1. BackupData
   2. BackupDataAll
   3. BackupDataStats
   4. CopyVolumeData
   5. DeleteData
   6. DeleteDataAll
   7. RestoreData

>    8. RestoreDataAll

Reviewer 4: As discussed, instead of adding versions, we would add new functions with the names having`*WithKopia` suffix.

Reviewer 4: @shlokc9 can you add a note about this is yet to decide or TBD.

Reviewer 6: Sure, I'll add a note about this. 👍🏼 
Maybe a separate design would be needed for handling kopia clients inside Kanister Functions....

   Resource called `RepositoryServer` as described in the following section.

#### Custom Resource Definition

This design proposes a new Custom Resource Definition (CRD) named
`RepositoryServer`, to represent a Kopia repository server.
It is a namespace-scoped resource, owned and managed by the controller in the
`kanister` namespace. As mentioned above, the CRD offers a set of parameters to
configure a Kopia repository server.

To limit the controller's RBAC scope, all the `RepositoryServer` resources
are created in the `kanister` namespace.

A sample `RepositoryServer` resource created to interact with the Kopia
repository of workloads running in the `monitoring` namespace looks like this:

```yaml
apiVersion: cr.kanister.io/v1alpha1
kind: RepositoryServer
metadata:
   name: repository-monitoring
   namespace: kanister
   labels:
      repo.kanister.io/target-namespace: monitoring
spec:
   storage:
      # required: storage location info
      secretRef:
         name: location
         namespace: kanister
      # required: creds to access the location
      # optional when using location type file-store
      credentialSecretRef:
         name: loc-creds

>          namespace: kanister

Reviewer 7: Is it safe to assume that these secrets are always going to be in the kanister (controller's) namespace. If yes, do we need the namespace field?

Reviewer 6: We have discussed this over a call with @pavannd1. This was intentionally added because we want the CR-controller to allow fetching secrets user create in their own namespace. This would be in-sync with the discussions to allow Blueprints or Profile CRs to be created in user/app namespace and not the controller(kanister) namespace

         namespace: kanister
      # optional: if specified, these values will be used by the controller to
      # override default username and hostname when connecting to the
      # repository from the server.
      # if not specified, the controller will use generated defaults
      username: kanisterAdmin
      hostname: monitoring
   server:
      # required: server admin details
      adminSecretRef:
         name: repository-monitoring-server-admin
         namespace: kanister
      # required: TLS certificate required for secure communication between the Kopia client and server
      tlsSecretRef:
         name: repository-monitoring-server-tls
         namespace: kanister
      # required: repository users list
      userAccessSecretRef:
         name: repository-monitoring-server-access
         namespace: kanister
      # required: selector for network policy required to enable cross-namespace access to the server
      networkPolicy:
         namespaceSelector:
            matchLabels:
               app: monitoring
         podSelector:
            matchLabels:
               pod: kopia-client
               app: monitoring

> status:

Reviewer 6: @pavannd1 We also discussed about adding a new string field `isReady` in the status subresource that would convey the users if RepositoryServer is ready to use or not
1. Value would either be "True" if ready
2. Or would be "False" if an error occurred in the controller

# Integrating Kopia with Kanister

This document proposes all the high-level changes required within Kanister to use [Kopia](https://kopia.io/) as the primary backup and restore tool.

## Problems

Kanister offers an in-house capability to perform backup and restore using some operation-specific Functions like BackupData, RestoreData, etc.
Although, they are much easier to use and significantly improve the user experience of Blueprint authors, these Functions have a lot of issues.

Few of those problems are;
1. They do not encrypt in-flight data during transfer
2. The backup data at-rest on the cloud stores are not secure
3. Operations are slow compared to other tools in the market

Therefore, they are not ideally recommended to be used for protecting application data.

> Instead, we make use of v2 Blueprints whose phases call a `kando` utility to perform data snapshots.
> This tool internally executes `Kopia` commands to manage backups in the cloud.
> 
> This approach is again not a robust solution because;
> 1. We have to create a separate kopia profile and use it within kando command to pass necessary kopia server info
> 2. We have to follow the correct kando command syntax in the Blueprint while backing up and mention `--output-name` in the Action outputArtifacts
> 3. Kopia errors are not logged when kando commands fail in the acion phase  

Reviewer: ```suggestion
Kanister also provides a command line utility `kando` that can be used to move data to and from object stores.
This tool internally executes `Kopia` commands to move the backup data.
The v2 version of the example Kanister Blueprints supports this. However, there are a few caveats to using these Blueprints.

1. `kando` uses `Kopia` only when a Kanister Profile of type `Kopia` is provided
2. A Kanister Profile of type `Kopia` requires a [Kopia Repository Server](https://kopia.io/docs/repository-server/) running in the same namespace as the Kanister controller
3. A Repository Server requires a [Kopia Repository](https://kopia.io/docs/repositories/) to be initialized on a backend storage target

Kanister currently lacks documentation and automation to use these features.
```

Author: Thanks @pavannd1 :) Your suggestions look better worded 👍🏼

## Introducing Kopia

Kopia is a powerful, cross-platform tool for managing encrypted backups in the cloud.
It provides fast and secure backups, using compression, data deduplication and client-side end-to-end encryption.
It supports a variety of backup storage targets, including object stores, which allows users to choose the storage provider that better addresses their needs.
It is a lock-free system that allows for concurrent multi-client operations including garbage collection.

To explore other features of Kopia, see its [documentation](https://kopia.io/docs/features/).

## Scope

1. Design the usage of [Kopia Repository Server](https://kopia.io/docs/repository-server/) as an separate on-demand workload
   - In order to make use of user access control and cloud storage credential abstraction offered in the server-based operations 
2. Re-work Kanister Functions like BackupData, RestoreData, CopyVolumeData, etc. with  Kopia workflows
   - And leverage the use of Kopia Repository Server workload within these Functions

### Design the usage of Kopia Repository Server

> 

Author: @pavannd1 Might need your help here to briefly state our CR/CRD design approach.

- Therefore, we plan to introduce a feature flag during Kanister installation that will decide if the Kopia Repository Server workload is supported or not

Proper steps to toggle between the Kopia Server workload support will be documented here later. By default, the feature flag would be disabled.

## User Experience

- Existing users and downstream consumers can continue to make use of previous functionality as per the above 'Backward Compatibility' section
- Which means upgrading the Kanister controller version wouldn't need any changes in existing blueprint workflows. 
- However, the user experience for new Kanister Functions is expected to change based on the design decisions for Kopia Repository Server workload usage
- Mainly, the user might have to perform CRUD on the Repository Server workload which would be an added prerequisite to work with Kopia-based Kanister Functions
- In case, the users switch off the feature flag for server workload and wish to create a server by themselves, 
  we aim to provide them with detailed steps to create a server workload as part of "bring-your-own-server" BYOS model

> ## Limitations to using Kopia API Server

Author: @pavannd1 Should we get rid of the "Limitations" section? Because these are quite vague right now. We will have clearer idea of limitations while working on the Kopia Repo Server CR design. And would be able to document it better on that doc.

Reviewer: This seems more like pre-requisites to using the server rather than limitations to me. We can reword accordingly
